Refactor to allow a basic setup without Haskell components

harendra-kumar opened the pull request:
The first commit is a refactor without changing any existing functionality. I have taken care that no behavior changes unless something might have crept in inadvertently.

Other than the refactor, it adds two command line options to install.sh. One for a basic installation without Haskell components and another to specify a repo to install from. I am providing the motivation for both below.

Basic Installation: haskell vim now has pretty good common (non-haskell) settings as well. I want to use the same setup even where I do not use haskell or in setups where installing stack and GHC would take too much space and time even though I won't need them. Now, I can just provide an option to install and have the same setup elsewhere as well. Even non-haskell users can now use haskell-vim-now, maybe we can call it vim-now instead!

Please note that it does not change or break anything for the existing end users.

repo path: I added this mainly for testing so that I could easily specify my git clone as the repo when running install.sh.

The second commit is to avoid confusion about the powerline fonts. I interpreted the message as the fonts have to be installed and configured though it was already installed only to be configured. I made that clear in the message.

SX91 commented:
I don't like the `setup_done` function. It's pretty useless. Maybe move it to the end of `setup`?

@harendra-kumar did you just split setup phases to separate functions? No setup sequence changes, no missing calls or checks?

harendra-kumar commented:
Thanks @SX91 !

1) I am sorry, I left the now dead setup() function around which might have caused confusion. I will push a commit to remove that.

2) It is mostly splitting with a little regrouping. If you see the diff fully expanded the major change you will see is line 95-134 moved to setup_haskell() function in the newer file. You can verify the sequence by comparing these two blocks. 

There has been minor change in sequence because of regrouping of similar functionality e.g.
* the stack exist check was the first thing in old setup() but now moved to setup_haskell()
* ctags exuberance check moved to setup_tools()
* git config modification now moved to setup_tools()

Other than that I think pretty much everything is intact. There should be no missing call. I have tested it on my setup and everything seems to be working fine till now.

3) Yeah, powerline-fonts is not being installed on Mac. But the original message still remains a bit confusing on other platforms. I have code to install powerline fonts on Mac as well. I can either add that or remove the changes I added. I will remove it for now. Maybe we can deal with it separately.

harendra-kumar commented:
I fixed, rebased and pushed. Removed:
* the powerline message changes
* removed the setup() function.

begriffs commented:
@harendra-kumar thanks for the pull request. Sorry it's been almost a month with no response since you submitted it.

This vim config is historically pretty opinionated and meant for Haskell. However after reviewing your changes I like how it separates the install process a little more cleanly. Your command line flag to use it as a non-Haskell config doesn't seem to affect Haskell users so it seems fine.

I'd like to merge it but I have two questions
* Can you rebase off master please? There was a change about how we deal with old vs new versions of Slack and this change now conflicts with your PR
* How gracefully will the installer handle a bogus value of `--repo`? Will it proceed and create a broken installation?
* Could you add a note in the readme about the new options?

harendra-kumar commented:
Thanks @begriffs! I completely understand the opinionated part and would not argue with that. The idea was to make it more modular and flexible without affecting the main focus. The changes are supposed to be helpful in general.

I will address your questions and push the changes for review again soon.

harendra-kumar commented:
1) rebased off the latest master
2) Made the --repo error handling more robust:
* If the setup script is not found in the repo we will exit. If it is a wrong repo and has a setup script at the expected place then we will fail when the correct functions are not found in the script. If it has correct functions as well then we cannot help.
* If the repo specified on the command line does not match the repo in the existing installation we will bail out.

3) Made the vim plugin install more robust against failures. We will not switch the config until plugins are installed successfully. Therefore preserving the existing config unless we have a new working config. This is something that I encountered when testing, so fixed this as well.

harendra-kumar commented:
Updated the readme. If it looks cluttered, maybe we can move the advanced install options to another page and just put a pointer to that in the readme? or move them  in some sort of appendix?

begriffs commented on README.md:
````diff
@@ -29,13 +29,40 @@ and lets you
 ```
 **WARNING**: this command, once successful, will _make backups and **remove**_ your existing VIM configurations (`.vim`, plugins etc). You can later [customize](#customizing) HVN configurations.
 
+### Basic Install
````
How about putting this at the end of the readme with a name like "Other Install Methods" or "Installing without Haskell"? It's good to have the info, but I'd like to keep the first paragraphs simple.

harendra-kumar replied:
Yeah, I think that would be better we should have the common case stuff concise enough.

begriffs commented on scripts/setup.sh:
```diff
@@ -14,34 +14,72 @@ stack_resolver() {
+
+  if ! check_exist stack >/dev/null ; then
+    err "Installer requires Stack."
+    msg "Installation instructions: https://github.com/commercialhaskell/stack#how-to-install"
```
While you're at it, this url has moved to http://docs.haskellstack.org/en/stable/README/#how-to-install

harendra-kumar commented:
I have updated the README and stack install URL, please take a look again.

begriffs commented:
Great work!

Before merging I'm investigating a strange message on Travis CI:

```
Error detected while processing /home/travis/.config/haskell-vim-now/.vimrc:
line   89:
E117: Unknown function: plug#begin
```

This message began appearing a few commits ago on this pull request. It doesn't make the build fail, but I think it probably should have.

To try debugging the problem I cleared Travis' build cache hoping a fresh install would be good, but now it's having even worse problems (probably unrelated to this pull request). https://travis-ci.org/begriffs/haskell-vim-now/builds/144133847

harendra-kumar commented:
Looks like a network problem, both of them. In the first one perhaps plug did not get installed (that's why plug#begin was not found) and in the second one the output from curl was perhaps empty or truncated (i.e. not having a proper gzip header).

begriffs commented:
OK the intermittent network problem went away but it is still issuing errors about installing things with vim-plug:

```
--- Checking ctags' exuberance...
--- Setting git to use fully-pathed vim for messages...
--- Installing vim-plug
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 66089  100 66089    0     0   203k      0 --:--:-- --:--:-- --:--:--  204k
--- Installing plugins using vim-plug...
Error detected while processing /home/travis/.config/haskell-vim-now/.vimrc:
line   89:
E117: Unknown function: plug#begin
line   93:
E492: Not an editor command: Plug 'jgdavey/tslime.vim'
line   94:
E492: Not an editor command: Plug 'Shougo/vimproc.vim', { 'do': 'make' }
line   95:
E492: Not an editor command: Plug 'ervandew/supertab'

-- etc etc
```

I re-ran the tests on the master branch and it installs the plugins fine, so I think this is specific to the current pull request. Can you look into it?

harendra-kumar commented:
In one of the later commits I changed the old vim config to be switched only after everything is installed successfully including plugins. But that did not work for plugins because they want the new .vim dir. It was working fine with my test setup because I had a stale .vim symlink already. Sorry for the trouble.

begriffs commented:
No trouble at all, thanks for your contribution.

begriffs commented:
If you're feeling adventurous you could file an issue on vim-plug about making the destination directory more configurable.
